MapboxDirections/MBSpeedLimit.swift

```diff
@@ -0,0 +1,112 @@
+//
+//  MBSpeedLimit.swift
+//  MapboxDirections
+//
+//  Created by Avi Cieplinski on 4/15/19.
+//  Copyright © 2019 Mapbox. All rights reserved.
+//
+
+import Foundation
+
+@objc(MBSpeedLimit)
+public class SpeedLimit: NSObject {
+
+    @objc public static let invalid = Double(-1)
+
+    /**
+     A unitless measure of speed which is dependent on the `SpeedLimit.speedUnit`.
+
+     By default, the speed will be unknown and equal to `MBSpeedIsInvalid` which is -1.
+
+     If the speed is none, the value will be equal to `Double.greatestFiniteMagnitude`.
+     */
+    @objc public var value: Double = invalid
+
+
+    /**
+     Units for `SpeedLimit.speed`.
+     */
+    @objc public var unit: SpeedUnit = .kilometersPerHour
+
+    /**
+     Initialize a new `SpeedLimit` object.
+     */
+    public init(value: Double, speedUnit: SpeedUnit) {
+        self.value = value
+        self.unit = speedUnit
+    }
+
+    /**
+     Initialize a new `SpeedLimit` object from a JSON dictionary.
+     */
+    @objc public convenience init(json: [String: Any]) {
+        var speed = json["speed"] as? Double ?? SpeedLimit.invalid
+        if let speedNone = json["unknown"] as? Bool, speedNone {
+            speed = .greatestFiniteMagnitude
+        }
+
+        let speedUnit: SpeedUnit
+        if let speedUnitString = json["unit"] as? String {
+            speedUnit = SpeedUnit(description: speedUnitString) ?? .kilometersPerHour
+        } else {
+            speedUnit = .kilometersPerHour
+        }
+
+        self.init(value: speed, speedUnit: speedUnit)
+    }
+
+    public static var supportsSecureCoding = true
+
+    public required init?(coder decoder: NSCoder) {
+        value = decoder.decodeDouble(forKey: "value")
+
+        guard let speedUnitString = decoder.decodeObject(of: NSString.self, forKey: "unit") as String?, let speedUnit = SpeedUnit(description: speedUnitString) else {
+            return nil
+        }
+        self.unit = speedUnit
+    }
+
+    public func encode(with coder: NSCoder) {
+        coder.encode(value, forKey: "value")
+        coder.encode(unit, forKey: "unit")
+    }
+}
+
+/**
+ Unit of measurement for `SpeedLimit`.
+ */
+@objc(MBSpeedUnit)
+public enum SpeedUnit: Int, CustomStringConvertible {
+
+    /**
+     Indicates the road segment uses miles per hour for measuring speed limits.
+     */
+    case milesPerHour
+
+    /**
+     Indicates the road segment uses kilometers per hour for measuring speed limits.
+     */
+    case kilometersPerHour
```

Comment:
Argh, it would be great if we could just use `Measurement` along with `UnitSpeed.kilometersPerHour`, but alas we still support iOS 9.0 (along with macOS 10.11.0, tvOS 9.0, and watchOS 2.0). I’m chomping at the bit to drop iOS 9 support.

/cc @mapbox/navigation-ios

Reply:
A proposed change to the Directions API would add information about locally relevant units to each step, making this class somewhat redundant. We could store the speed limit annotations as `CLLocationSpeed`s (concerting everything to meters per second). Then client code could losslessly convert back to kilometers per hour or miles per hour as indicated by the step.